### PR TITLE
chore: add ADR for OSI token

### PR DESCRIPTION
**Pull request by bloodearnest**
Fixes: #2636

### CODE REVIEW

**Review by lucyb — APPROVED** [review]
This is really helpful. Thank you!

File: docs/adr/0014-osi-token.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+## Status
+Accepted
+
+## Context
+
+For OSI, job-server needs to be able to write generated code to Github.
+Previously, job-server only had read access to repos, as a security control.
+Granting arbitrary repo write permissions to job-server weakens the security
+layering, as it increases the possibly attcak vectors of what compromising
```
Comment (ghickman) [inline docs/adr/0014-osi-token.md:12]:
```suggestion
layering, as it increases the possible attack vectors of what compromising
```

File: docs/adr/0014-osi-token.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+job-server alone can do.
+
+The current read-only access is provided by the opensafely-readonly bot
+account, as classic PATs can not be restricted to read only access in any other
```
Comment (ghickman) [inline docs/adr/0014-osi-token.md:16]:
```suggestion
account, as classic PATs cannot be restricted to read only access in any other
```

**Review by ghickman — APPROVED** [review]
Bunch of nitty tweaks but this is great 👍

File: docs/adr/0014-osi-token.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# 14. Github token set up for OSI
+Date: 2023-03-24
+
+## Status
+Accepted
+
+## Context
+
+For OSI, job-server needs to be able to write generated code to Github.
+Previously, job-server only had read access to repos, as a security control.
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply docs/adr/0014-osi-token.md:10]: I mean, yes, but the design is that is was readonly. The fact that there's a loop hole I don't think is really relevant to this decision or consequences.

And fixing that loop hole will also probably require the secure service.

File: docs/adr/0014-osi-token.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+## Consequences
+
+Job-server does now have a token than can write to repos. It is at least
+restricted to writing to interactive repos, and cannot access user's repos
+
+This unblocks OSIv2, but it does however remove the layer of protection we had
+originaly designed.  A compromised job-server can now run arbitrary code in any
+backend. We still have our other layers of protection (Level 4 access
+controls), so this is not critical.
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply docs/adr/0014-osi-token.md:44]: Compromising job-server does bypass our output checking process, as we use job-server to know who is allowed to release.

File: docs/adr/0014-osi-token.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bloodearnest) [reply docs/adr/0014-osi-token.md:10]: There's a small window where *write access to a brand new repo* is possible, for clarity. Only repo creation is the problem.

**Pull request merged**